src/misc/umap.c

```diff
@@ -0,0 +1,262 @@
+        value = VECTOR(*values)[i];
+        if (igraph_vector_int_size(smallest) < k) {
+            igraph_vector_int_push_back(smallest, i);
+            igraph_vector_push_back(&smallest_values, i);
+        } else {
+            max_index = 0;
+            for (igraph_integer_t j = 1; j < igraph_vector_int_size(smallest); j++) {
+                if (VECTOR(smallest_values)[j] > VECTOR(smallest_values)[max_index]) {
+                    max_index = j;
+                }
+            }
+            if (value > VECTOR(smallest_values)[max_index]) {
+                VECTOR(smallest_values)[max_index] = value;
+                VECTOR(*smallest)[max_index] = i;
+            }
+        }
+    }
+    igraph_vector_destroy(&smallest_values);
+    IGRAPH_FINALLY_CLEAN(1);
+    return (IGRAPH_SUCCESS);
+}
+
+/*This creates a k-nearest neighbors graph with their distances.*/
+/*first trying everything with simple slow algorithms */
+static igraph_error_t igraph_knng_from_euclid(igraph_matrix_t *data,
+        igraph_t *knn_graph, igraph_vector_t *knn_distances) {
+    igraph_integer_t k = 10;
+    igraph_integer_t no_of_nodes = igraph_matrix_nrow(data);
+    igraph_vector_t node_distances;
+    igraph_vector_t data_row_i;
+    igraph_vector_t data_row_j;
+    igraph_vector_int_t closest;
+    igraph_vector_int_t edges;
+
+    IGRAPH_VECTOR_INIT_FINALLY(&node_distances, no_of_nodes);
+    IGRAPH_VECTOR_INIT_FINALLY(&data_row_i, 0);
+    IGRAPH_VECTOR_INIT_FINALLY(&data_row_j, 0);
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&closest, k);
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&edges, 0);
+    for (igraph_integer_t i = 0; i < no_of_nodes; i++) {
+        igraph_matrix_get_row(data, &data_row_i, i);
+        for (igraph_integer_t j = i; j < no_of_nodes; j++) {
+            igraph_matrix_get_row(data, &data_row_j, j);
+            igraph_vector_sub(&data_row_j, &data_row_i);
+            VECTOR(node_distances)[j] = igraph_vector_sumsq(&data_row_j);
+        }
+        igraph_get_k_smallest(&node_distances, &closest, k, i); //uses i as a starting point. Very ugly interface
+        for (igraph_integer_t j = 0; j < igraph_vector_int_size(&closest); j++) { //size is usually k, but not for last few nodes
+            IGRAPH_CHECK(igraph_vector_int_push_back(&edges, i));
+            IGRAPH_CHECK(igraph_vector_int_push_back(&edges, VECTOR(closest)[j]));
+            IGRAPH_CHECK(igraph_vector_push_back(knn_distances, VECTOR(node_distances)[VECTOR(closest)[j]]));
+        }
+    }
+    IGRAPH_CHECK(igraph_create(knn_graph, &edges, no_of_nodes, IGRAPH_UNDIRECTED));
+    igraph_vector_destroy(&node_distances);
+    igraph_vector_destroy(&data_row_i);
+    igraph_vector_destroy(&data_row_j);
+    igraph_vector_int_destroy(&closest);
+    igraph_vector_int_destroy(&edges);
+    IGRAPH_FINALLY_CLEAN(5);
+    return (IGRAPH_SUCCESS);
+}
+
+/*open set size is just the size of the distance to the closest neighbor*/
```

Comment:
notice that the concept of *distance* here is a little misleading. Physicists find it useful, but in practice it's usually defined as something like

    d (v1, v2) := 1 - correlation(v1, v2)

which is actually funny because then it's this weird object between 0 and 2. In other words, the edge weight as a *similarity* score, something like max(0, correlation(v1, v2)) is a better representation of the situation compared to rephrasing the whole problem as a distance matrix.

Of course, the embedding space itself is a 2D Euclidean thing, so distances matter there. But that does not change the fact that in the original space, similarity/edge weight has more meaning than distance.

Not sure how useful all of this actually is for you now ;-)